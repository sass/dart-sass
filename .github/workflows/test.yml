name: Test

defaults:
  run: {shell: bash}

on:
  workflow_call:
  workflow_dispatch:

jobs:
  format:
    name: Code formatting
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v4
    - uses: dart-lang/setup-dart@v1
    - run: dart format --fix .
    - run: git diff --exit-code

  static_analysis:
    name: Static analysis
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with: {github-token: "${{ github.token }}"}

      - name: Analyze Dart
        run: dart analyze --fatal-warnings ./

  dartdoc:
    name: Dartdoc
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with: {github-token: "${{ github.token }}"}

      - name: dartdoc sass
        run: dart run dartdoc --quiet --no-generate-docs
          --errors ambiguous-doc-reference,broken-link,deprecated
          --errors unknown-directive,unknown-macro,unresolved-doc-reference
      - name: dartdoc sass_api
        run: cd pkg/sass_api && dart run dartdoc --quiet --no-generate-docs
          --errors ambiguous-doc-reference,broken-link,deprecated
          --errors unknown-directive,unknown-macro,unresolved-doc-reference

  sass_spec_language:
    name: "Language Tests | Dart ${{ matrix.dart_channel }} | ${{ matrix.async_label }}"
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        dart_channel: [stable, dev]
        async_label: [synchronous]
        async_args: ['']
        include:
          - dart_channel: stable
            async_label: asynchronous
            async_args: '--cmd-args --async'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
      - uses: ./.github/util/sass-spec

      - name: Run specs
        run: npm run sass-spec -- --dart .. $extra_args
        working-directory: sass-spec
        env: {extra_args: "${{ matrix.async_args }}"}

  sass_spec_js:
    name: "JS API Tests | Pure JS | Dart ${{ matrix.dart_channel }} | Node ${{ matrix.node-version }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3
          # Test LTS version with dart dev channel
          - os: ubuntu-latest
            dart_channel: dev
            node-version: 'lts/*'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}
      - uses: ./.github/util/sass-spec

      - name: Build JS
        run: dart run grinder pkg-npm-dev

      - name: Check out Sass specification
        uses: sass/clone-linked-repo@v1
        with:
          repo: sass/sass
          path: language

      - name: Run tests
        run: npm run js-api-spec -- --sassSassRepo ../language --sassPackage ../build/npm
        working-directory: sass-spec

  sass_spec_js_embedded:
    name: 'JS API Tests | Embedded | Node ${{ matrix.node-version }} | ${{ matrix.os }}'
    runs-on: ${{ matrix.os }}
    if: "github.event_name != 'pull_request' || !contains(github.event.pull_request.body, 'skip sass-embedded')"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}
      - uses: ./.github/util/sass-spec

      - name: Check out the embedded host
        uses: sass/clone-linked-repo@v1
        with: {repo: sass/embedded-host-node}

      - name: Check out the language repo
        uses: sass/clone-linked-repo@v1
        with: {repo: sass/sass, path: build/language}

      - name: Initialize embedded host
        run: |
          npm install
          npm run init -- --compiler-path=.. --language-path=../build/language
          npm run compile
          mv {`pwd`/,dist/}lib/src/vendor/dart-sass
        working-directory: embedded-host-node

      - name: Version info
        run: |
          path=embedded-host-node/dist/lib/src/vendor/dart-sass/sass
          if [[ -f "$path.cmd" ]]; then "./$path.cmd" --version
          elif [[ -f "$path.bat" ]]; then "./$path.bat" --version
          elif [[ -f "$path.exe" ]]; then "./$path.exe" --version
          else "./$path" --version
          fi

      - name: Run tests
        run: npm run js-api-spec -- --sassPackage ../embedded-host-node --sassSassRepo ../build/language
        working-directory: sass-spec

  sass_spec_js_browser:
    name: "JS API Tests | Browser | Dart ${{ matrix.dart_channel }}"

    strategy:
      matrix:
        dart_channel: [stable]
      fail-fast: false

    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: browser-actions/setup-chrome@v1
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
      - uses: ./.github/util/sass-spec

      - name: Build JS
        run: dart run grinder pkg-npm-dev

      - name: Install built dependencies
        run: npm install
        working-directory: build/npm

      - name: Check out Sass specification
        uses: sass/clone-linked-repo@v1
        with:
          repo: sass/sass
          path: language

      - name: Run tests
        run: npm run js-api-spec -- --sassSassRepo ../language --sassPackage ../build/npm --browser
        working-directory: sass-spec
        env:
          CHROME_EXECUTABLE: chrome

  dart_tests:
    name: "Dart tests | Dart ${{ matrix.dart_channel }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        # TODO(nweiz): Re-enable this when
        # https://github.com/dart-lang/sdk/issues/52121#issuecomment-1728534228
        # is addressed.
        # include: [{os: ubuntu-latest, dart_channel: dev}]

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}

      - run: dart run grinder pkg-standalone-dev
      - name: Run tests
        run: dart run test -x node

  # Unit tests that use Node.js, defined in test/.
  node_tests:
    name: "Node tests | Dart ${{ matrix.dart_channel }} | Node ${{ matrix.node-version }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false

      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3
          # Test LTS version with dart dev channel
          - os: ubuntu-latest
            dart_channel: dev
            node-version: 'lts/*'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}

      - run: dart run grinder pkg-npm-dev
      - name: Run tests
        run: dart run test -t node -j 2

  browser_tests:
    name: "Browser Tests | Dart ${{ matrix.dart_channel }}"

    strategy:
      matrix:
        dart_channel: [stable]
      fail-fast: false

    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - uses: browser-actions/setup-chrome@v1
    - uses: ./.github/util/initialize
      with:
        dart-sdk: ${{ matrix.dart_channel }}
        github-token: ${{ github.token }}

    - run: dart run grinder pkg-npm-dev
    - name: Run tests
      run: dart run test -p chrome -j 2
      env:
        CHROME_EXECUTABLE: chrome
