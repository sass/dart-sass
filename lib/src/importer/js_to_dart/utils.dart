// Copyright 2023 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:node_interop/js.dart';

import '../../js/utils.dart';
import '../utils.dart';

/// Throws a JsError if [scheme] isn't a valid URL scheme.
void validateUrlScheme(String scheme) {
  if (!isValidUrlScheme(scheme)) {
    jsThrow(
        JsError('"$scheme" isn\'t a valid URL scheme (for example "file").'));
  }
}
