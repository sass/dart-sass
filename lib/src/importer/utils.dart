// Copyright 2017 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:path/path.dart' as p;

import '../io.dart';

/// Whether the Sass compiler is currently evaluating an `@import` rule.
///
/// When evaluating `@import` rules, URLs should canonicalize to an import-only
/// file if one exists for the URL being canonicalized. Otherwise,
/// canonicalization should be identical for `@import` and `@use` rules. It's
/// admittedly hacky to set this globally, but `@import` will eventually be
/// removed, at which point we can delete this and have one consistent behavior.
bool _inImportRule = false;

/// Runs [callback] in a context where [resolveImportPath] uses `@import`
/// semantics rather than `@use` semantics.
T inImportRule<T>(T callback()) {
  var wasInImportRule = _inImportRule;
  _inImportRule = true;
  try {
    return callback();
  } finally {
    _inImportRule = wasInImportRule;
  }
}

/// Like [inImportRule], but asynchronous.
Future<T> inImportRuleAsync<T>(Future<T> callback()) async {
  var wasInImportRule = _inImportRule;
  _inImportRule = true;
  try {
    return await callback();
  } finally {
    _inImportRule = wasInImportRule;
  }
}

/// Resolves an imported path using the same logic as the filesystem importer.
///
/// This tries to fill in extensions and partial prefixes and check for a
/// directory default. If no file can be found, it returns `null`.
String resolveImportPath(String path) {
  var extension = p.extension(path);
  if (extension == '.sass' || extension == '.scss' || extension == '.css') {
    return _ifInImport(() => _exactlyOne(
            _tryPath('${p.withoutExtension(path)}.import$extension'))) ??
        _exactlyOne(_tryPath(path));
  }

  return _ifInImport(
          () => _exactlyOne(_tryPathWithExtensions('$path.import'))) ??
      _exactlyOne(_tryPathWithExtensions(path)) ??
      _tryPathAsDirectory(path);
}

/// Like [_tryPath], but checks `.sass`, `.scss`, and `.css` extensions.
List<String> _tryPathWithExtensions(String path) {
  var result = _tryPath(path + '.sass')..addAll(_tryPath(path + '.scss'));
  return result.isNotEmpty ? result : _tryPath(path + '.css');
}

/// Returns the [path] and/or the partial with the same name, if either or both
/// exists.
///
/// If neither exists, returns an empty list.
List<String> _tryPath(String path) {
  var partial = p.join(p.dirname(path), "_${p.basename(path)}");
  return [if (fileExists(partial)) partial, if (fileExists(path)) path];
}

/// Returns the resolved index file for [path] if [path] is a directory and the
/// index file exists.
///
/// Otherwise, returns `null`.
String _tryPathAsDirectory(String path) {
  if (!dirExists(path)) return null;

  return _ifInImport(() =>
          _exactlyOne(_tryPathWithExtensions(p.join(path, 'index.import')))) ??
      _exactlyOne(_tryPathWithExtensions(p.join(path, 'index')));
}

/// If [paths] contains exactly one path, returns that path.
///
/// If it contains no paths, returns `null`. If it contains more than one,
/// throws an exception.
String _exactlyOne(List<String> paths) {
  if (paths.isEmpty) return null;
  if (paths.length == 1) return paths.first;

  throw "It's not clear which file to import. Found:\n" +
      paths.map((path) => "  " + p.prettyUri(p.toUri(path))).join("\n");
}

/// If [_inImportRule] is `true`, invokes callback and returns the result.
///
/// Otherwise, returns `null`.
T _ifInImport<T>(T callback()) => _inImportRule ? callback() : null;
