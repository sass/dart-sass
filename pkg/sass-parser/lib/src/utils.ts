// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import * as postcss from 'postcss';

import {Node} from './node';

/**
 * A type that matches any constructor for {@link T}. From
 * https://www.typescriptlang.org/docs/handbook/mixins.html.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Constructor<T> = new (...args: any[]) => T;

/**
 * An explicit field description passed to `cloneNode` that describes in detail
 * how to clone it.
 */
interface ExplicitClonableField<Name extends string> {
  /** The field's name. */
  name: Name;

  /**
   * Whether the field can be set to an explicit undefined value which means
   * something different than an absent field.
   */
  explicitUndefined?: boolean;
}

/** The type of field names that can be passed into `cloneNode`. */
type ClonableField<Name extends string> = Name | ExplicitClonableField<Name>;

/** Makes a {@link ClonableField} explicit. */
function parseClonableField<Name extends string>(
  field: ClonableField<Name>
): ExplicitClonableField<Name> {
  return typeof field === 'string' ? {name: field} : field;
}

/**
 * Creates a copy of {@link node} by passing all the properties in {@link
 * constructorFields} as an object to its constructor.
 *
 * If {@link overrides} is passed, it overrides any existing constructor field
 * values. It's also used to assign {@link assignedFields} after the cloned
 * object has been constructed.
 */
export function cloneNode<T extends Pick<postcss.Node, 'source'>>(
  node: T,
  overrides: Record<string, unknown> | undefined,
  constructorFields: ClonableField<keyof T & string>[],
  assignedFields?: ClonableField<keyof T & string>[]
): T {
  // We have to do these casts because the actual `...Prop` types that get
  // passed in and used for the constructor aren't actually subtypes of
  // `Partial<T>`. They use `never` types to ensure that various properties are
  // mutually exclusive, which is not compatible.
  const typedOverrides = overrides as Partial<T> | undefined;
  const constructorFn = node.constructor as new (defaults: Partial<T>) => T;

  const constructorParams: Partial<T> = {};
  for (const field of constructorFields) {
    const {name, explicitUndefined} = parseClonableField(field);
    let value: T[keyof T & string] | undefined;
    if (
      typedOverrides &&
      (explicitUndefined
        ? Object.hasOwn(typedOverrides, name)
        : typedOverrides[name] !== undefined)
    ) {
      value = typedOverrides[name];
    } else {
      value = maybeClone(node[name]);
    }
    if (value !== undefined) constructorParams[name] = value;
  }
  const cloned = new constructorFn(constructorParams);

  if (typedOverrides && assignedFields) {
    for (const field of assignedFields) {
      const {name, explicitUndefined} = parseClonableField(field);
      if (
        explicitUndefined
          ? Object.hasOwn(typedOverrides, name)
          : typedOverrides[name]
      ) {
        // This isn't actually guaranteed to be non-null, but TypeScript
        // (correctly) complains that we could be passing an undefined value to
        // a field that doesn't allow undefined. We don't have a good way of
        // forbidding that while still allowing users to override values that do
        // explicitly allow undefined, though.
        cloned[name] = typedOverrides[name]!;
      }
    }
  }

  cloned.source = node.source;
  return cloned;
}

/**
 * If {@link value} is a Sass node, a record, or an array, clones it and returns
 * the clone. Otherwise, returns it as-is.
 */
function maybeClone<T>(value: T): T {
  if (Array.isArray(value)) return value.map(maybeClone) as T;
  if (typeof value !== 'object' || value === null) return value;
  // The only records we care about are raws, which only contain primitives and
  // arrays of primitives, so structued cloning is safe.
  if (value.constructor === Object) return structuredClone(value);
  if (value instanceof postcss.Node) return value.clone() as T;
  return value;
}

/**
 * Converts {@link node} into a JSON-safe object, with the given {@link fields}
 * included.
 *
 * This always includes the `type`, `sassType`, `raws`, and `source` fields if
 * set. It converts multiple references to the same source input object into
 * indexes into a top-level list.
 */
export function toJSON<T extends Node>(
  node: T,
  fields: (keyof T & string)[],
  inputs?: Map<postcss.Input, number>
): object {
  // Only include the inputs field at the top level.
  const includeInputs = !inputs;
  inputs ??= new Map();
  let inputIndex = inputs.size;

  const result: Record<string, unknown> = {};
  if ('type' in node) result.type = (node as {type: string}).type;

  fields = ['sassType', 'raws', ...fields];
  for (const field of fields) {
    const value = node[field];
    if (value !== undefined) result[field] = toJsonField(field, value, inputs);
  }

  if (node.source) {
    let inputId = inputs.get(node.source.input);
    if (inputId === undefined) {
      inputId = inputIndex++;
      inputs.set(node.source.input, inputId);
    }

    result.source = {
      start: node.source.start,
      end: node.source.end,
      inputId,
    };
  }

  if (includeInputs) {
    result.inputs = [...inputs.keys()].map(input => input.toJSON());
  }
  return result;
}

/**
 * Converts a single field with name {@link field} and value {@link value} to a
 * JSON-safe object.
 *
 * The {@link inputs} map works the same as it does in {@link toJSON}.
 */
function toJsonField(
  field: string,
  value: unknown,
  inputs: Map<postcss.Input, number>
): unknown {
  if (typeof value !== 'object' || value === null) {
    return value;
  } else if (Array.isArray(value)) {
    return value.map((element, i) =>
      toJsonField(i.toString(), element, inputs)
    );
  } else if ('toJSON' in value) {
    if ('sassType' in value) {
      return (
        value as {
          toJSON: (field: string, inputs: Map<postcss.Input, number>) => object;
        }
      ).toJSON('', inputs);
    } else {
      return (value as {toJSON: (field: string) => object}).toJSON(field);
    }
  } else {
    return value;
  }
}

/**
 * Returns the longest string (of code units) that's an initial substring of
 * every string in
 * {@link strings}.
 */
export function longestCommonInitialSubstring(strings: string[]): string {
  let candidate: string | undefined;
  for (const string of strings) {
    if (candidate === undefined) {
      candidate = string;
    } else {
      for (let i = 0; i < candidate.length && i < string.length; i++) {
        if (candidate.charCodeAt(i) !== string.charCodeAt(i)) {
          candidate = candidate.substring(0, i);
          break;
        }
      }
      candidate = candidate.substring(
        0,
        Math.min(candidate.length, string.length)
      );
    }
  }
  return candidate ?? '';
}
