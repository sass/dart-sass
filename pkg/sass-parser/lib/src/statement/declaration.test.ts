// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import {
  Declaration,
  Interpolation,
  Rule,
  StringExpression,
  sass,
  scss,
} from '../..';
import * as utils from '../../../test/utils';

describe('a property declaration', () => {
  let node: Declaration;
  beforeEach(
    () =>
      void (node = new Declaration({
        prop: 'foo',
        expression: {text: 'bar'},
      })),
  );

  describe('with no children', () => {
    function describeNode(
      description: string,
      create: () => Declaration,
    ): void {
      describe(description, () => {
        beforeEach(() => (node = create()));

        it('has a type', () => expect(node.type.toString()).toBe('decl'));

        it('has a sassType', () =>
          expect(node.sassType.toString()).toBe('decl'));

        it('has propInterpolation', () =>
          expect(node).toHaveInterpolation('propInterpolation', 'foo'));

        it('has a prop', () => expect(node.prop).toBe('foo'));

        it('has an expression', () =>
          expect(node).toHaveStringExpression('expression', 'bar'));

        it('has a value', () => expect(node.value).toBe('bar'));

        it('has no nodes', () => expect(node.nodes).toBeUndefined());
      });
    }

    describeNode(
      'parsed as SCSS',
      () =>
        (scss.parse('a {foo: bar}').nodes[0] as Rule).nodes[0] as Declaration,
    );

    describeNode(
      'parsed as Sass',
      () =>
        (sass.parse('a\n  foo: bar').nodes[0] as Rule).nodes[0] as Declaration,
    );

    describe('constructed manually', () => {
      describeNode(
        'with prop and value',
        () =>
          new Declaration({
            prop: 'foo',
            value: 'bar',
          }),
      );

      describe('with propInterpolation', () => {
        describeNode(
          "that's a string",
          () =>
            new Declaration({
              propInterpolation: 'foo',
              value: 'bar',
            }),
        );

        describeNode(
          "that's child props",
          () =>
            new Declaration({
              propInterpolation: {nodes: ['foo']},
              value: 'bar',
            }),
        );

        describeNode(
          "that's an explicit Interpolation",
          () =>
            new Declaration({
              propInterpolation: new Interpolation('foo'),
              value: 'bar',
            }),
        );
      });

      describe('with an expression', () => {
        describeNode(
          "that's child props",
          () =>
            new Declaration({
              prop: 'foo',
              expression: {text: 'bar'},
            }),
        );

        describeNode(
          "that's an Expression",
          () =>
            new Declaration({
              prop: 'foo',
              expression: new StringExpression({text: 'bar'}),
            }),
        );
      });
    });

    describeNode('constructed from ChildProps', () =>
      utils.fromChildProps({prop: 'foo', value: 'bar'}),
    );
  });

  describe('with a value and children', () => {
    function describeNode(
      description: string,
      create: () => Declaration,
    ): void {
      describe(description, () => {
        beforeEach(() => (node = create()));

        it('has a type', () => expect(node.type.toString()).toBe('decl'));

        it('has a sassType', () =>
          expect(node.sassType.toString()).toBe('decl'));

        it('has propInterpolation', () =>
          expect(node).toHaveInterpolation('propInterpolation', 'foo'));

        it('has a prop', () => expect(node.prop).toBe('foo'));

        it('has an expression', () =>
          expect(node).toHaveStringExpression('expression', 'bar'));

        it('has a value', () => expect(node.value).toBe('bar'));

        it('has nodes', () => expect(node.nodes).toHaveLength(1));
      });
    }

    describeNode(
      'parsed as SCSS',
      () =>
        (scss.parse('a {foo: bar {baz: bang}}').nodes[0] as Rule)
          .nodes[0] as Declaration,
    );

    describeNode(
      'parsed as Sass',
      () =>
        (sass.parse('a\n  foo: bar\n    baz: bang').nodes[0] as Rule)
          .nodes[0] as Declaration,
    );

    describeNode(
      'constructed manually',
      () =>
        new Declaration({
          prop: 'foo',
          value: 'bar',
          nodes: [{name: 'baz'}],
        }),
    );

    describeNode('constructed from ChildProps', () =>
      utils.fromChildProps({prop: 'foo', value: 'bar', nodes: [{name: 'baz'}]}),
    );
  });

  it('assigned a new prop', () => {
    node.prop = 'baz';
    expect(node.prop).toBe('baz');
    expect(node).toHaveInterpolation('propInterpolation', 'baz');
  });

  it('assigned a new propInterpolation', () => {
    node.propInterpolation = 'baz';
    expect(node.prop).toBe('baz');
    expect(node).toHaveInterpolation('propInterpolation', 'baz');
  });

  it('assigned a new expression', () => {
    const old = node.expression!;
    node.expression = {text: 'baz'};
    expect(old.parent).toBeUndefined();
    expect(node).toHaveStringExpression('expression', 'baz');
    expect(node.value).toBe('baz');
  });

  it('assigned a value', () => {
    node.value = 'Helvetica, sans-serif';
    expect(node).toHaveStringExpression('expression', 'Helvetica, sans-serif');
    expect(node.value).toBe('Helvetica, sans-serif');
  });

  it('is not a variable without --', () => expect(node.variable).toBe(false));

  it('is a variable with --', () => {
    node.prop = '--foo';
    expect(node.variable).toBe(true);
  });

  describe('stringifies', () => {
    describe('to SCSS', () => {
      describe('with default raws', () => {
        it('with no children', () =>
          expect(new Declaration({prop: 'foo', value: 'bar'}).toString()).toBe(
            'foo: bar',
          ));

        it('with a value and children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              nodes: [{prop: 'baz', value: 'bang'}],
            }).toString(),
          ).toBe('foo: bar {\n    baz: bang\n}'));

        it('with only children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              nodes: [{prop: 'baz', value: 'bang'}],
            }).toString(),
          ).toBe('foo: {\n    baz: bang\n}'));
      });

      describe('with before', () => {
        it('on its own', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              raws: {before: '/**/'},
            }).toString(),
          ).toBe('foo: bar'));

        it('as a child', () =>
          expect(
            new Rule({
              selector: 'foo',
              nodes: [{prop: 'bar', value: 'baz', raws: {before: '/**/'}}],
            }).toString(),
          ).toBe('foo {/**/bar: baz\n}'));
      });

      describe('with between', () => {
        it('with no children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              raws: {between: ':/**/'},
            }).toString(),
          ).toBe('foo:/**/bar'));

        it('with a value and children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              nodes: [{prop: 'baz', value: 'bang'}],
              raws: {between: ':/**/'},
            }).toString(),
          ).toBe('foo:/**/bar {\n    baz: bang\n}'));

        it('with only children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              nodes: [{prop: 'baz', value: 'bang'}],
              raws: {between: ':/**/'},
            }).toString(),
          ).toBe('foo:/**/ {\n    baz: bang\n}'));
      });

      it('ignores important', () =>
        expect(
          new Declaration({
            prop: 'foo',
            value: 'bar !important',
            raws: {important: '!IMPORTANT'},
          }).toString(),
        ).toBe('foo: bar !important'));

      it('ignores value', () =>
        expect(
          new Declaration({
            prop: 'foo',
            value: 'bar',
            raws: {value: {value: 'bar', raw: 'BAR'}},
          }).toString(),
        ).toBe('foo: bar'));

      describe('with afterValue', () => {
        it('with no children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              raws: {afterValue: '/**/'},
            }).toString(),
          ).toBe('foo: bar/**/'));

        it('with a value and children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              nodes: [{prop: 'baz', value: 'bang'}],
              raws: {afterValue: '/**/'},
            }).toString(),
          ).toBe('foo: bar/**/{\n    baz: bang\n}'));

        it('with no value and children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              nodes: [{prop: 'baz', value: 'bang'}],
              raws: {afterValue: '/**/'},
            }).toString(),
          ).toBe('foo:/**/{\n    baz: bang\n}'));
      });

      describe('with after', () => {
        it('without children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              raws: {after: '/**/'},
            }).toString(),
          ).toBe('foo: bar'));

        it('with children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              nodes: [{prop: 'baz', value: 'bang'}],
              raws: {after: '/**/'},
            }).toString(),
          ).toBe('foo: bar {\n    baz: bang/**/}'));
      });

      describe('with ownSemicolon', () => {
        it('without children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              raws: {ownSemicolon: ';'},
            }).toString(),
          ).toBe('foo: bar'));

        it('with children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              nodes: [{prop: 'baz', value: 'bang'}],
              raws: {ownSemicolon: ';'},
            }).toString(),
          ).toBe('foo: bar {\n    baz: bang\n};'));
      });

      describe('with semicolon', () => {
        it('without children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              raws: {semicolon: true},
            }).toString(),
          ).toBe('foo: bar'));

        it('with children', () =>
          expect(
            new Declaration({
              prop: 'foo',
              value: 'bar',
              nodes: [{prop: 'baz', value: 'bang'}],
              raws: {semicolon: true},
            }).toString(),
          ).toBe('foo: bar {\n    baz: bang;\n}'));
      });
    });
  });

  describe('clone', () => {
    let original: Declaration;
    beforeEach(() => {
      original = (scss.parse('a {foo: bar {baz: bang}}').nodes[0] as Rule)
        .nodes[0] as Declaration;
      // TODO: remove this once raws are properly parsed
      original.raws.between = ' :';
    });

    describe('with no overrides', () => {
      let clone: Declaration;
      beforeEach(() => void (clone = original.clone()));

      describe('has the same properties:', () => {
        it('prop', () => expect(clone.prop).toBe('foo'));

        it('propInterpolation', () =>
          expect(clone).toHaveInterpolation('propInterpolation', 'foo'));

        it('expression', () =>
          expect(clone).toHaveStringExpression('expression', 'bar'));
      });

      describe('creates a new', () => {
        it('self', () => expect(clone).not.toBe(original));

        for (const attr of [
          'propInterpolation',
          'expression',
          'raws',
        ] as const) {
          it(attr, () => expect(clone[attr]).not.toBe(original[attr]));
        }
      });
    });

    describe('overrides', () => {
      describe('raws', () => {
        it('defined', () =>
          expect(original.clone({raws: {afterValue: '  '}}).raws).toEqual({
            afterValue: '  ',
          }));

        it('undefined', () =>
          expect(original.clone({raws: undefined}).raws).toEqual({
            between: ' :',
          }));
      });

      describe('propInterpolation', () => {
        describe('defined', () => {
          let clone: Declaration;
          beforeEach(() => {
            clone = original.clone({propInterpolation: 'zap'});
          });

          it('changes variableName', () =>
            expect(clone).toHaveInterpolation('propInterpolation', 'zap'));

          it('changes prop', () => expect(clone.prop).toBe('zap'));
        });

        describe('undefined', () => {
          let clone: Declaration;
          beforeEach(() => {
            clone = original.clone({propInterpolation: undefined});
          });

          it('preserves variableName', () =>
            expect(clone).toHaveInterpolation('propInterpolation', 'foo'));

          it('preserves prop', () => expect(clone.prop).toBe('foo'));
        });
      });

      describe('expression', () => {
        it('defined changes expression', () =>
          expect(
            original.clone({expression: {text: 'zap'}}),
          ).toHaveStringExpression('expression', 'zap'));

        it('undefined removes expression', () =>
          expect(
            original.clone({expression: undefined}).expression,
          ).toBeUndefined());
      });
    });
  });

  describe('toJSON', () => {
    it('with expression and nodes', () =>
      expect(
        (scss.parse('a {foo: bar {baz: bang}}').nodes[0] as Rule).nodes[0],
      ).toMatchSnapshot());

    it('with expression and no nodes', () =>
      expect(
        (scss.parse('a {foo: bar}').nodes[0] as Rule).nodes[0],
      ).toMatchSnapshot());

    it('with no expression and nodes', () =>
      expect(
        (scss.parse('a {foo: {baz: bang}}').nodes[0] as Rule).nodes[0],
      ).toMatchSnapshot());
  });
});
