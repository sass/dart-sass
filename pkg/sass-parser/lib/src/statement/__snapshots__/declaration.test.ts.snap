// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a property declaration toJSON with expression and no nodes 1`] = `
{
  "expression": <bar>,
  "inputs": [
    {
      "css": "a {foo: bar}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "prop": "foo",
  "propInterpolation": <foo>,
  "raws": {},
  "sassType": "decl",
  "source": <1:4-1:12 in 0>,
  "type": "decl",
  "value": "bar",
}
`;

exports[`a property declaration toJSON with expression and nodes 1`] = `
{
  "expression": <bar>,
  "inputs": [
    {
      "css": "a {foo: bar {baz: bang}}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "nodes": [
    <baz: bang>,
  ],
  "prop": "foo",
  "propInterpolation": <foo>,
  "raws": {},
  "sassType": "decl",
  "source": <1:4-1:24 in 0>,
  "type": "decl",
  "value": "bar",
}
`;

exports[`a property declaration toJSON with no expression and nodes 1`] = `
{
  "inputs": [
    {
      "css": "a {foo: {baz: bang}}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "nodes": [
    <baz: bang>,
  ],
  "prop": "foo",
  "propInterpolation": <foo>,
  "raws": {},
  "sassType": "decl",
  "source": <1:4-1:20 in 0>,
  "type": "decl",
  "value": "",
}
`;
